e2e test script

bbguimaraes opened the pull request:
End-to-end testing script.  Details on how to run it on the README.

There are some initial post-deploy validations used mainly to test if the
script is working, but it might be worth getting them in anyway.

bbguimaraes committed:
E2E testing script.

bbguimaraes committed:
Some post-deployment validations.

bbguimaraes committed:
Option to use local deployer image.

bbguimaraes committed:
Now using origin-integration-common scripts.

sosiouxme commented on hack/testing/e2e_deployment.sh:
```diff
@@ -0,0 +1,162 @@
+set -euo pipefail
+
+ORIGIN_APIMAN_ROOT=$(realpath --no-symlinks "$(dirname "$BASH_SOURCE")/../..")
+OS_ROOT=${OS_ROOT:-$(realpath --no-symlinks "$ORIGIN_APIMAN_ROOT/../origin")}
```
I was trying to get rid of this requirement for having origin cloned nearby, or relying on it at all. It looks to me like the only need for OS_ROOT at this point is to include the openshift binaries in the PATH. Would it be unreasonable to just assume these will be in your path for this script?

sosiouxme commented on deployer/scripts/validate.sh:
```diff
@@ -67,3 +63,100 @@ function check_routes() {
+			truststore.password
+		apiman-console \
+			ca.crt client.crt client.key client.keystore client.keystore.password \
+			keystore keystore.password truststore truststore.password
```
insert "gateway.user" at the beginning of this line and the corresponding one in apiman-gateway

sosiouxme commented on deployer/scripts/validate.sh:
```diff
@@ -67,3 +63,100 @@ function check_routes() {
+  # 52: The server didn't reply anything.
+  local deadline ret
+  deadline=$(($(date +%s) + 3 * 60))
+  while [ "$(date +%s)" -lt "$deadline" ]; do
```
we should probably have our own "wait_for" function in common sort of like in origin's test scripts. it's probably a little awkward to do that in this environment though where you want to show the user something contextual.

sosiouxme commented on deployer/scripts/validate.sh:
```diff
@@ -67,3 +63,100 @@ function check_routes() {
+		services apiman-console apiman-es-cluster apiman-gateway apiman-storage
+		routes apiman-console apiman-gateway
+		EOF
+}
```
maybe you just haven't gotten to it yet, but i was hoping to also see validations that the pods actually started and became endpoints, like with metrics https://github.com/openshift/origin-metrics/blob/56d168af4d39c4bd6c0195bac4016b868350ce92/deployer/scripts/validate.sh#L232

sosiouxme replied:
also, checking that the routes have been accepted by a router etc https://github.com/openshift/origin-metrics/blob/56d168af4d39c4bd6c0195bac4016b868350ce92/deployer/scripts/validate.sh#L444

sosiouxme commented on hack/testing/e2e_deployment.sh:
```diff
@@ -0,0 +1,162 @@
+BUILD_IMAGES=${BUILD_IMAGES:-}
+LOCAL_IMAGE=${LOCAL_IMAGE:-}
+CLEANUP=${CLEANUP:-}
+TMP_PROJECT=${TMP_PROJECT:-}
```
seems like you could just use ${:-} below and save a little awkwardness here?

sosiouxme commented on hack/testing/e2e_deployment.sh:
```diff
@@ -0,0 +1,162 @@
+    fi
+    if [ "$TMP_PROJECT" ]; then
+        TMP_PROJECT=apiman-e2e-test-$( \
+            base64 /dev/urandom | tr -cd [a-z0-9] | head -c 5 || true)
```
if you just want 5 chars then $(mktemp -u XXXXX) will do

sosiouxme commented on hack/testing/e2e_deployment.sh:
```diff
@@ -0,0 +1,162 @@
+    os::cmd::expect_success \
+        'oc label secret/apiman-deployer apiman-infra=deployer'
+    [ "$build_images" ] \
+        && filter="jq '{apiVersion,kind,items:[.items[]|del(.spec.triggers)]}'"
```
Can you document any requirements like jq in the README? It exists in the devenv but we need to make sure it's not hard to figure out what you need anywhere else.

sosiouxme commented on hack/testing/e2e_deployment.sh:
```diff
@@ -0,0 +1,162 @@
+    os::cmd::expect_success \
+        "$(echo \
+            oc process -f '"$ORIGIN_APIMAN_ROOT/hack/dev-builds.yaml"' \
+            \| $filter \| oc create -f -)"
```
alternative to jq, you could maybe use oc patch if sufficiently clever (note, it creates instead of having output)

sosiouxme commented on hack/testing/e2e_deployment.sh:
```diff
@@ -0,0 +1,162 @@
+        'oc secrets new apiman-deployer nothing=/dev/null'
+    os::cmd::expect_success \
+        'oc label secret/apiman-deployer apiman-infra=deployer'
+    [ "$build_images" ] \
```
shouldn't this just go in build_images() ? I may be missing something.

sosiouxme commented on hack/testing/e2e_deployment.sh:
```diff
@@ -0,0 +1,162 @@
+        "$get_is_tags origin" '^latest$' "$((2 * TIME_MIN ))" 10
+    os::cmd::try_until_text \
+        "$get_is_tags centos" '^7$' "$((2 * TIME_MIN ))" 10
+    for x in deployer elasticsearch curator; do
```
we can add now: builder console gateway, except the last two need to build from a different repo (https://github.com/fabric8io/fabric8-ipaas/), which I guess we can assume is cloned beside this repo.

sosiouxme commented on hack/testing/README:
```diff
@@ -0,0 +1,35 @@
+environment variables.  The common could also be executed this way:
+
+    BOOTSTRAP_OS=1 DEPLOY=1 BUILD_IMAGES \
+        hack/testing/e2e_deployment.sh --tmp-project
```
Great README, could you mention any non-obvious dependencies like jq, what should be checked out where, any other potential setup steps?

sosiouxme commented on hack/testing/e2e_deployment.sh:
```diff
@@ -0,0 +1,162 @@
+    image_prefix=${2:+-v IMAGE_PREFIX="$( \
+        oc get imagestream/apiman-elasticsearch \
+            --template '{{.status.dockerImageRepository}}' \
+                | sed 's,[^/]*$,,')"}
```
could we get the image prefix before changing to --tmp-project? assuming i have everything nice and built the way i like, then want to test deployment in a throwaway project...

bbguimaraes committed:
Minor fixes.

bbguimaraes committed:
Remove the need for the origin repo.

bbguimaraes committed:
Rework of the option for image reuse.

bbguimaraes commented:
@sosiouxme: updates addressing your review.  Some comments:
- `hack/dev-local-builds.yaml` was created based on `hack/dev-builds.yaml` to
  avoid having to do a lot of patching on the code and needing `jq`.
- The way images are built/pulled was reworked.  I think it is much better (or
  at least less hack-y).  The names of some flags were changed because of that,
  the `README` was updated and has a few examples.

It's still based on an older version, I'll rebase and update.

bbguimaraes committed:
Fix for local source builds.

Images were mistakenly built from source *and* from git.

bbguimaraes committed:
Updated to build gateway/console images.

bbguimaraes committed:
Small improvement for temporary projects.

bbguimaraes commented:
@sosiouxme PTAL

Updated to the latest master, all components should be deploying and running
correctly.

**edit**: no workaround needed anymore

We currently have a small problem on the console, but the correction should be
ready soon.  In the meantime you can use this workaround to test it:

``` diff
diff --git a/deployer/scripts/install.sh b/deployer/scripts/install.sh
index 0777b8c..40958bb 100644
--- a/deployer/scripts/install.sh
+++ b/deployer/scripts/install.sh
@@ -92,6 +92,10 @@ function create_config() {

   # use or generate server certs
   local file component hostnames secret domain=${project}.svc.cluster.local
+  cat > $scratch_dir/apiman.properties <<EOF
+es.username=es.username
+es.password=$(printf es.password | base64)
+EOF
   for component in console gateway elasticsearch curator; do
     if [ -s $secret_dir/apiman-${component}.keystore.jks ]; then
       # use files from secret when present
@@ -128,6 +132,8 @@ function create_config() {
           client.keystore.password=$scratch_dir/${user}.keystore.jks.password
           ca.crt=$scratch_dir/ca.crt
           )
+    [ "$component" == console ] \
+      && secret+=( apiman.properties=$scratch_dir/apiman.properties )
     case "$component" in
       console|gateway) secret+=( gateway.user=$scratch_dir/gateway.user ) ;;
       elasticsearch)   secret+=( searchguard-node-key=$scratch_dir/searchguard-node-key.key ) ;;
diff --git a/deployer/scripts/validate.sh b/deployer/scripts/validate.sh
index a540607..91b3263 100644
--- a/deployer/scripts/validate.sh
+++ b/deployer/scripts/validate.sh
@@ -86,6 +86,7 @@ function check_secrets() {
            keystore keystore.password searchguard-node-key truststore \
            truststore.password
        apiman-console \
+           apiman.properties \
            ca.crt client.crt client.key client.keystore client.keystore.password \
            gateway.user keystore keystore.password truststore truststore.password
        apiman-gateway \
```
